### PR TITLE
[MXNET-240] Fix CUDA compile error

### PR DESCRIPTION
**Pull request by zheng-da**
## Description ##
Fix remaining compile errors with only CUDA enabled.

This disable CuDNN in CentOS CI to test compilation with CUDA only.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

**Comment by zheng-da** [comment]
@marcoabreu This one is to address your PR https://github.com/apache/incubator-mxnet/pull/10281

**Comment by zheng-da** [comment]
It seems we can't run tests on cuda only.

```
======================================================================

ERROR: test_operator_gpu.test_lstm_cpu_inference

----------------------------------------------------------------------

Traceback (most recent call last):

  File "/usr/lib/python3.6/site-packages/nose/case.py", line 198, in runTest

    self.test(*self.arg)

  File "/usr/lib/python3.6/site-packages/nose/util.py", line 620, in newfunc

    return func(*arg, **kw)

  File "/work/mxnet/tests/python/gpu/../unittest/test_gluon_rnn.py", line 79, in test_lstm_cpu_inference

    y = model(x).asnumpy()

  File "/work/mxnet/python/mxnet/gluon/block.py", line 361, in __call__

    return self.forward(*args)

  File "/work/mxnet/python/mxnet/gluon/rnn/rnn_layer.py", line 191, in forward

    out = self._forward_kernel(inputs, states)

  File "/work/mxnet/python/mxnet/gluon/rnn/rnn_layer.py", line 225, in _forward_kernel

    p=self._dropout, state_outputs=True, mode=self._mode)

  File "<string>", line 73, in RNN

  File "/work/mxnet/python/mxnet/_ctypes/ndarray.py", line 92, in _imperative_invoke

    ctypes.byref(out_stypes)))

  File "/work/mxnet/python/mxnet/base.py", line 149, in check_call

    raise MXNetError(py_str(_LIB.MXGetLastError()))

mxnet.base.MXNetError: [20:35:37] src/operator/rnn.cu:43: RNN is only available for cuDNN at the moment.



Stack trace returned 10 entries:

[bt] (0) /work/mxnet/python/mxnet/../../lib/libmxnet.so(dmlc::StackTrace()+0x42) [0x7f87cfb7db92]

[bt] (1) /work/mxnet/python/mxnet/../../lib/libmxnet.so(dmlc::LogMessageFatal::~LogMessageFatal()+0x18) [0x7f87cfb7e1f8]

[bt] (2) /work/mxnet/python/mxnet/../../lib/libmxnet.so(mxnet::Operator* mxnet::op::CreateOp<mshadow::gpu>(mxnet::op::RNNParam, int)+0x3c) [0x7f87d404bfec]

[bt] (3) /work/mxnet/python/mxnet/../../lib/libmxnet.so(mxnet::op::RNNProp::CreateOperatorEx(mxnet::Context, std::vector<nnvm::TShape, std::allocator<nnvm::TShape> >*, std::vector<int, std::allocator<int> >*) const+0x55) [0x7f87d2515415]

[bt] (4) /work/mxnet/python/mxnet/../../lib/libmxnet.so(mxnet::op::OpPropCreateLayerOp(nnvm::NodeAttrs const&, mxnet::Context, std::vector<nnvm::TShape, std::allocator<nnvm::TShape> > const&, std::vector<int, std::allocator<int> > const&)+0x268) [0x7f87d2282c48]

[bt] (5) /work/mxnet/python/mxnet/../../lib/libmxnet.so(std::_Function_handler<mxnet::OpStatePtr (nnvm::NodeAttrs const&, mxnet::Context, std::vector<nnvm::TShape, std::allocator<nnvm::TShape> > const&, std::vector<int, std::allocator<int> > const&), mxnet::OpStatePtr (*)(nnvm::NodeAttrs const&, mxnet::Context, std::vector<nnvm::TShape, std::allocator<nnvm::TShape> > const&, std::vector<int, std::allocator<int> > const&)>::_M_invoke(std::_Any_data const&, nnvm::NodeAttrs const&, mxnet::Context, std::vector<nnvm::TShape, std::allocator<nnvm::TShape> > const&, std::vector<int, std::allocator<int> > const&)+0x15) [0x7f87cfcf15e5]

[bt] (6) /work/mxnet/python/mxnet/../../lib/libmxnet.so(mxnet::Imperative::InvokeOp(mxnet::Context const&, nnvm::NodeAttrs const&, std::vector<mxnet::NDArray*, std::allocator<mxnet::NDArray*> > const&, std::vector<mxnet::NDArray*, std::allocator<mxnet::NDArray*> > const&, std::vector<mxnet::OpReqType, std::allocator<mxnet::OpReqType> > const&, mxnet::DispatchMode, mxnet::OpStatePtr)+0xaa5) [0x7f87d22bfba5]

[bt] (7) /work/mxnet/python/mxnet/../../lib/libmxnet.so(mxnet::Imperative::Invoke(mxnet::Context const&, nnvm::NodeAttrs const&, std::vector<mxnet::NDArray*, std::allocator<mxnet::NDArray*> > const&, std::vector<mxnet::NDArray*, std::allocator<mxnet::NDArray*> > const&)+0x4ec) [0x7f87d22c026c]

[bt] (8) /work/mxnet/python/mxnet/../../lib/libmxnet.so(MXImperativeInvokeImpl(void*, int, void**, int*, void***, int, char const**, char const**)+0x4aa) [0x7f87d27c827a]

[bt] (9) /work/mxnet/python/mxnet/../../lib/libmxnet.so(MXImperativeInvokeEx+0x6f) [0x7f87d27c882f]


```

**Comment by marcoabreu** [comment]
Thank you, Da! What do you think about making those tests conditional in terms of skipping them if CuDNN is not available?

**Comment by zheng-da** [comment]
I'm not sure. It seems difficult.
maybe let's fix the compilation error first and worry about running the test later.

**Comment by marcoabreu** [comment]
Well there's no immediate hurry as long as this PR makes it into the release, so I think you can do as you prefer.

Is there no way to determine on the python side whether CuDNN is present or not?

**Comment by zheng-da** [comment]
i don't know if there is a way.

**Comment by marcoabreu** [comment]
@szha @piiswrong could you assist?

File: src/operator/nn/batch_norm.cu [hunk 1]
```diff
@@ -708,7 +708,7 @@ void BatchNormGradCompute<gpu>(const nnvm::NodeAttrs& attrs,
     })
   }
 #else
-  MSHADOW_REAL_TYPE_SWITCH_EX(out_grad[0].type_flag_, DType, AccReal, {
+  MSHADOW_REAL_TYPE_SWITCH_EX(dtype, DType, AccReal, {
```
Comment (marcoabreu) [inline src/operator/nn/batch_norm.cu]: @tdomhan please have a look here

Comment (tdomhan) [reply src/operator/nn/batch_norm.cu]: I guess the dtype of the inputs and outputs must always be the same. Using the dtype variable (as done in this PR) seems cleaner.

**Comment by marcoabreu** [comment]
Issue is tracked at https://github.com/apache/incubator-mxnet/issues/10353

**Pull request merged**